### PR TITLE
[WIP] Promotion Engine Enhancements

### PR DESCRIPTION
**Pull request by peteward**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | yes |
| Deprecations? | no |
| Fixed tickets | #3368 |
| License | MIT |

Implements #3368. Please read for a deeper understanding of what this is trying to achieve.

**It's WIP so don't review code style. Tests will break since @pjedrzejewski merged Driver refactor this morning**

I haven't had as much time as I wanted today to polish this. Despite the incomplete tasks, a large amount of the hard work has been done.
- [x] New core structure of `Actions` which are now groupings of `Benefits` and `Filters`
- [x] Implementation of several base Filters - Taxon, Archetype, Cheapest, Most Expensive, Every, Price
- [x] Many new tests added and existing updated. Core behat tests have been implemented to decouple UI to test domain logic.
- [x] FixedDiscount features file needs work (will do this tonight).
- [ ] `Action` probably needs a `position` to handle priority.
- [ ] Admin UI needs implementing as per the referenced issue. 
- [ ] `PromotionApplicator` needs decoupling as it's currently implemented to expect Orders (`getItems()`), for the purposes of filtering order items within it. Possibly a `getFilterSubjects()` or `getFilterable()` method? 
- [ ] Similarly deciding what should be a PromotionSubject - just an `Order`? 
- [ ] Behat tests need an overhaul in structure. There are too many duplicates spread in different files. There should be features to cover admin UI, some for frontend UI, then the main bulk to cover the permutations of available promotion building.
- [ ] Integration tests to be removed? These were from before @pamil made behat fast.
- [ ] Finally the actual application of Adjustments resulting from Benefits needs to be thought out. Currently it's working at the OrderItem-level, which works but breaks some of the current UI tests. A plan should be worked out how to complete this in a complimentary way to changes suggested in #3712.

Suggested next steps:
- Defined the changes to the Admin UI via a replacement Behat file.
- Implement Admin UI to populate implemented structure.
- The rest!

### CODE REVIEW

**Comment by peteward** [comment]
The main result of these changes are:

* Allow an administrator to 'filter' the contents of the cart to only apply a promotional benefit to those that match it.
* To improve upon the Exclusivity feature which allows you to either use one exclusive promotion on it's own or any non-exclusive promotion in combination. These changes will allow specific benefits to be combined through a single promotion.

To illustrate, here are some example scenarios. The bold ones can be achieved with the new design only:

* Get 20% off your first order
* Get 20% off if spend more than £200
* Get a free gift if you buy at least one dress
* **Get a free gift for every dress you buy**
* **Get a 10% off jacket if you buy trousers**
* Get free shipping if you spend more than 500
* **Get a free gift card and 20% off dresses if you buy dress**
* **Get 10% off any sale dresses**
* **Get £10 off suits if you spend more than 500 and get free shipping**
* **Get 20% off suits if you spend more than 500 and get £50 off order**
* **Get 10% off bow tie and belts if you buy jackets or trousers**
* **Get 10% off on dresses if these costs more than 300 and get £15 off jackets if these costs more than 250 when you place first order**

File: src/Sylius/Component/Promotion/Applicator/PromotionApplicator.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+use Sylius\Component\Promotion\Model\PromotionSubjectInterface;
+use Sylius\Component\Registry\ServiceRegistryInterface;
+
+/**
+ * Applies all registered promotion actions to given subject.
+ *
+ * TODO: This needs to consider coupling. It's written at the Component level, but expecting items.
+ *
+ * @author Saša Stamenković <umpirsky@gmail.com>
+ * @author  Piotr Walków <walkow.piotr@gmail.com>
```
Comment (tristanbes) [inline src/Sylius/Component/Promotion/Applicator/PromotionApplicator.php]: extra space after author annotation

File: src/Sylius/Component/Promotion/Applicator/PromotionApplicator.php [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    /**
+     * @param PromotionSubjectInterface $subject
+     * @param PromotionInterface $promotion
+     */
+    public function revert(PromotionSubjectInterface $subject, PromotionInterface $promotion)
+    {
+        foreach ($promotion->getActions() as $action) {
+            foreach ($action->getBenefits() as $benefit) {
+                $this->benefitRegistry
```
Comment (tristanbes) [inline src/Sylius/Component/Promotion/Applicator/PromotionApplicator.php]: Should the CS be like 
```
$this
    ->benefitRegistry
    ->....
;
```

File: src/Sylius/Component/Promotion/Filter/ArchetypeFilter.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    protected function resolveConfiguration(array $configuration)
+    {
+        $resolver = new OptionsResolver();
+
+        $resolver->setRequired(self::OPTION_ARCHETYPE);
+        $resolver->setAllowedTypes(self::OPTION_ARCHETYPE, 'int');
+
+        return $resolver->resolve($configuration);
+    }
+}
```
Comment (tristanbes) [inline src/Sylius/Component/Promotion/Filter/ArchetypeFilter.php]: some files has EOF new line, and some have not in this PR. 
Does Sylius recommend EOF new line or not in their CS guide ?

File: src/Sylius/Component/Promotion/Test/Filter/ArchetypeFilterTest.php [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+        $expected = new ArrayCollection([
+            $item1, $item2,
+        ]);
+
+        $this->assertCount(2, $result);
+        $this->assertEquals($expected, $result);
+    }
+
+    ###################################################################
```
Comment (tristanbes) [inline src/Sylius/Component/Promotion/Test/Filter/ArchetypeFilterTest.php]: are those special delimiters ? or should it be removed ?

File: src/Sylius/Component/Promotion/Test/Filter/CheapestFilterTest.php [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    public function it_returns_element_with_the_least_unit_price($givenCollection, $expectedItem)
+    {
+        $result = $this->sut->apply($givenCollection);
+
+        $this->assertInstanceOf(ArrayCollection::class, $result);
+        $this->assertCount(1, $result);
+        $this->assertSame($expectedItem, $result->first());
+    }
+
+    ########################################################################
```
Comment (tristanbes) [inline src/Sylius/Component/Promotion/Test/Filter/CheapestFilterTest.php]: same question as above :)

**Comment by peteward** [comment]
This PR is WIP and not ready for CS review yet (as per the line in bold above :)) 

Someone from the Lakion team will be finishing it over the coming weeks.

**Comment by tristanbes** [comment]
oh ok,I admit, I read the new promotions rules, but not the begining of the PR :D

**Comment by Mr-Negative** [comment]
Whats the status of this PR? any chances of finishing it? Becouse i see that 

> "Someone from the Lakion team will be finishing it over the coming weeks."

 don't happened ":-/

**Comment by Mr-Negative** [comment]
Whats the status of this PR? any chances of finishing it? Becouse i see that 

> "Someone from the Lakion team will be finishing it over the coming weeks."

 don't happened ":-/

**Comment by pamil** [comment]
Parts of this PR has been already implemented, being 2 years old in pre-stable era is quite much :)

**Pull request closed without merging**